### PR TITLE
Issue 189 | Add WebsocketProvider

### PR DESCRIPTION
**Pull request by ryan-rowland**
#189 

Happy path implementation. Specific errors and error codes haven't been observed or considered. A cheat sheet on which errors to expect and how to handle them would be helpful.

Works against test networks. Mainnet seems to want to disconnect the websocket almost immediately on each connection, causing some data loss. We should hold off on merging until we figure out why this is.

Test by running `node example.js`, which runs both `FetchProvider` and `WebsocketProvider` side by side to compare results.

### CODE REVIEW

**Comment by ryan-rowland** [comment]
@danfinlay Looks like there may be an issue with `wss://mainnet.infura.io/ws`. All of the other endpoints, including `wss://mainnet.infura.io/_ws`, work fine. Could I get a review based on the assumption the main endpoint is broken and will be updated?

PS: looks like circleCI is currently broken for this repo, causing a few approved PRs to not be merged.

File: util/create-payload.js [hunk 1]
```diff
@@ -4,11 +4,10 @@ const extend = require('xtend')
 module.exports = createPayload
 
 
-function createPayload(data){
+function createPayload(data) {
   return extend({
     // defaults
     id: getRandomId(),
-    jsonrpc: '2.0',
```
Comment (kumavis) [inline util/create-payload.js]: does the ws endpoint not accept this key?

**Comment by kumavis** [comment]
Added the websocket provider from this pull request, thank you

**Comment by lazaridiscom** [comment]
@kumavis , I tried to spot the relevant commit, but failed. Which is it?

**Comment by ryan-rowland** [comment]
@lazaridiscom Looks like it's here https://github.com/MetaMask/provider-engine/commit/586134128754a3bf1f31d76208a94dff168bf60b

And there are a few follow-up commits making some adjustments.

Thanks for the update @kumavis

File: util/create-payload.js (hunk 1 above)
Comment (ryan-rowland) [reply util/create-payload.js]: It does, but looking at the RPC over WebSocket spec I saw this was missing from requests and figured it better to be exact.

**Pull request closed without merging**